pkg/fakedata/generator.go

```diff
@@ -95,126 +77,141 @@ var date = func(column Column) string {
+}
+
+func customInteger(options string) (func() string, error) {
 	min := 0
 	max := 1000
+	var low, high string
+	intRange := strings.Split(options, ",")
+	low = intRange[0]
 
-	var _min, _max string
-	rng := strings.Split(column.Options, ",")
-	_min = rng[0]
-
-	if len(rng) > 1 {
-		_max = rng[1]
+	if len(intRange) > 1 {
+		high = intRange[1]
 	}
 
-	if len(_min) > 0 {
-		m, err := strconv.Atoi(_min)
+	if len(low) > 0 {
+		m, err := strconv.Atoi(low)
 		if err != nil {
-			fmt.Printf("could not convert min: %v", err)
-			os.Exit(1)
+			return nil, fmt.Errorf("could not convert min: %v", err)
 		}
 
 		min = m
 
-		if len(_max) > 0 {
-			m, err := strconv.Atoi(_max)
+		if len(high) > 0 {
+			m, err := strconv.Atoi(high)
 			if err != nil {
-				fmt.Printf("could not convert max: %v", err)
-				os.Exit(1)
+				return nil, fmt.Errorf("could not convert max: %v", err)
 			}
 
 			max = m
 		}
 	}
 
 	if min > max {
-		fmt.Printf("max(%d) is smaller than min(%d) in %v", max, min, column)
-		os.Exit(1)
+		return nil, fmt.Errorf("max(%d) is smaller than min(%d)", max, min)
 	}
 
-	return strconv.Itoa(min + rand.Intn(max+1-min))
+	return func() string { return _integer(min, max) }, nil
 }
 
-var enum = func(column Column) string {
-	enum := []string{"foo", "bar", "baz"}
+func file(path string) (func() string, error) {
+	if len(path) == 0 {
+		return nil, fmt.Errorf("no file path given")
+	}
+
+	filePath := strings.Trim(path, "'\"")
 
-	if len(column.Options) > 1 {
-		enum = strings.Split(column.Options, ",")
+	content, err := ioutil.ReadFile(filePath)
+	if err != nil {
+		return nil, fmt.Errorf("could not read file %s: %v", filePath, err)
 	}
+	list := strings.Split(string(content), "\n")
 
-	return withEnum(enum)(column)
+	return func() string { return withList(list)() }, nil
 }
 
-var fileCache map[string][]string
+type factory struct {
+	generators map[string]Generator
+}
 
-var file = func(column Column) string {
-	constraint := column.Options
-	if len(constraint) == 0 {
-		fmt.Printf("%s: no file path given", column.Name)
-		os.Exit(1)
+func (f factory) getGenerator(key, options string) (gen Generator, err error) {
+	gen, ok := f.generators[key]
+	if !ok {
+		return gen, fmt.Errorf("unknown generator: %s", key)
 	}
 
-	filePath := strings.Trim(constraint, "'\"")
+	customFn := gen.Func
 
-	list, ok := fileCache[filePath]
-	if !ok {
-		content, err := ioutil.ReadFile(filePath)
-		if err != nil {
-			fmt.Printf("could not read file %s: %v", filePath, err)
-			os.Exit(1)
+	switch key {
+	case "int":
+		customFn, err = customInteger(options)
+	case "date":
+		customFn, err = customDate(options)
+	case "enum":
+		list := []string{"foo", "bar", "baz"}
+		if len(options) > 0 {
+			list = strings.Split(options, ",")
 		}
-		list = strings.Split(string(content), "\n")
+		customFn = func() string { return withList(list)() }
+	case "file":
+		customFn, err = file(options)
 	}
 
-	return list[rand.Intn(len(list))]
+	gen.Func = customFn
+
+	return gen, err
 }
```

Comment:
The diff doesn't help here. But `getGenerator` (which I'm thinking of renaming into `newGenerator` but always not sure about naming) is the core of the change.

We fetch the generator from the factory and the `switch` does all the work for the "generators with customOptions. I think over time we can make this even "more obvious" (using interfaces and structs) but it's good enough for me now that we have only 3/4 generators with "custom needs"